README.md
![PyChord](./pychord.png)

# PyChord ![Build Status](https://github.com/yuma-m/pychord/actions/workflows/build.yml/badge.svg) [![Documentation Status](https://readthedocs.org/projects/pychord/badge/?version=latest)](http://pychord.readthedocs.io/en/latest/?badge=latest)

## Overview

Pychord is a python library to handle musical chords.

## Installation

```sh
$ pip install pychord
```

## Basic Usage

### Create a Chord

```python
>>> from pychord import Chord
>>> c = Chord("Am7")
>>> c
<Chord: Am7>
>>> c.info()
"""
Am7
root=A
quality=m7
appended=[]
on=None
"""
```

### Transpose a Chord

```python
>>> c = Chord("Am7/G")
>>> c.transpose(3)
>>> c
<Chord: Cm7/Bb>
```

### Get component notes

```python
>>> c = Chord("Am7")
>>> c.components()
['A', 'C', 'E', 'G']
>>> c.components_with_pitch(root_pitch=3)
['A3', 'C4', 'E4', 'G4']
```

### Compare Chords

```python
>>> Chord("C") == Chord("D")
False
>>> Chord("C#") == Chord("Db")
True
>>> c = Chord("C")
>>> c.transpose(2)
>>> c == Chord("D")
True
```

### Find Chords from notes

```python
>>> from pychord import find_chords_from_notes
>>> find_chords_from_notes(["C", "E", "G"])
[ <Chord: C>]
>>> find_chords_from_notes(["F#", "A", "C", "D"])
[ <Chord: D7/F#>]
>>> find_chords_from_notes(["F", "G", "C"])
[ <Chord: Fsus2>, <Chord: Csus4/F>]
```

### Create and handle chord progressions

```python
>>> from pychord import ChordProgression
>>> cp = ChordProgression(["C", "G/B", "Am"])
>>> cp
<ChordProgression: C | G/B | Am>

>>> cp.append("Em/G")
>>> cp
<ChordProgression: C | G/B | Am | Em/G>

>>> cp.transpose(+3)
>>> cp
<ChordProgression: Eb | Bb/D | Cm | Gm/Bb>

>>> cp[1]
<Chord: Bb/D>
```

## Advanced Usage

### Create a Chord from note index in a scale

```python
>>> Chord.from_note_index(note=1, quality="", scale="Cmaj")
<Chord: C>  # I of C major
>>> Chord.from_note_index(note=3, quality="m7", scale="Fmaj")
<Chord: Am7>  # IIIm7 of F major
>>> Chord.from_note_index(note=5, quality="7", scale="Amin")
<Chord: E7>  # V7 of A minor
```

### Overwrite the default Quality components with yours

```python
>>> from pychord import Chord, QualityManager
>>> Chord("C11").components()
['C', 'G', 'Bb', 'D', 'F']

>>> quality_manager = QualityManager()
>>> quality_manager.set_quality("11", (0, 4, 7, 10, 14, 17))
>>> Chord("C11").components()
['C', 'E', 'G', 'Bb', 'D', 'F']
```

### Inversions

Chord inversions are created with a forward slash and a number
indicating the order. This can optionally be combined with an
additional forward slash to change the bass note:

```python
>>> Chord("C/1").components() # First inversion of C
['E', 'G', 'C']
>>> Chord("C/2").components() # Second inversion of C
['G', 'C', 'E']

>>> Chord("Cm7/3/F").components() # Third inversion of Cm7 with an added F bass
['F', 'Bb', 'C', 'Eb', 'G']
```

## Examples

- [pychord-midi.py](./examples/pychord-midi.py) - Create a MIDI file using PyChord and pretty_midi.

## Supported Python Versions

- 3.6 and above

Python 2.7 and 3.5 compatibility was dropped from version 1.0.0.

## Links

- PyPI: https://pypi.python.org/pypi/pychord
- GitHub: https://github.com/yuma-m/pychord
- Documentation: http://pychord.readthedocs.io/en/latest/

## Author

- [Yuma Mihira](https://yuma.cloud/)

## License

- MIT License

Icon is made by [Freepik](https://www.flaticon.com/authors/freepik).


CHANGELOG.md
## Forthcoming

## v1.2.2

- Add tests for "M" and "maj" qualities synonyms
- Add missing synonyms for M qualities
- Contributor: @nickurak

## v1.2.1

- Add `get_qualities` method to `QualityManager`.
- Contributor: @nickurak

## v1.2.0

- Add `chromatic` parameter for `from_note_index`.
- Add `7(b9)`, `6b5` qualities.
- Contributor: @Moustov, @hejops

## v1.1.1

- Add `m11` and `maj13` qualities.
- Fix TypeError.
- Contributor: @Moustov

## v1.1.0

- Add `no5`, `m(no5)`, `(b5)`, `sus4add2` and `sus4add9` qualities.
- Support inversions.
- Contributor: @EnigmaCurry

## v1.0.0

- Drop compatibility for Python 2.7 and 3.5.
- Refactor whole library to optimize for Python 3.x.
  - Add type hints.
  - Use f-strings.
  - Rename `note_to_chord` to `find_chords_from_notes`.
  - Use tuple instead of list for `Quality.components`.

## v0.6.3

- Add modes to `RELATIVE_KEY_DICT`.
- Add `diatonic` option for `from_note_index`.
- Contributor: @hejops

## v0.6.2

- Add `add4`, `add11` and `add13` quality variations.
- Contributor: @philipmat

## v0.6.1

- Fix `dim6` to `dim7`.
- Add `M7+11` and `m7+5` qualities.

## v0.6.0

- Add an example to create a MIDI file.
- Add `QualityManager` class to overwrite default qualities.
  - Do not import `QUALITY_DICT` from modules other than quality.
- Fix `sus` quality.

## v0.5.1

- Add `m7b9b5` quality.

## v0.5.0

- Add Chord.components_with_pitch method.

## v0.4.2

- Rename `6/9` to `69`.
- Add `m69`, `m7b5` and `-` qualities.
- Contributor: @dok

## v0.4.1

- Add qualities.
  - `sus`, `maj7`, `maj9`, `m6`, `madd9`
  - `7b5`, `7#5`, `7b9`, `7#9`
  - `9-5`, `9b5`, `9+5`, `9#5`
  - `7#9b5`, `7#9#5`, `7b9b5`, `7b9#5`, `7#11`, `7b9#9`, `7b9#11`, `7#9#11`, `7b13`
  - `7b9b13`, `9+11`, `9#11`, `13-9`, `13b9`, `13+9`, `13#9`, `13+11`, `13#11`
- Support quality alias comparison.
  - `maj7 == M7`

## v0.4.0

- Add Chord.from_note_index method.
  - Support chord creation using note index in a scale.
- Contributor: @kwadwo00

## v0.3.2

- Make `QUALITY_DICT` values immutable.

## v0.3.1

- Raise TypeError in `__eq__` methods.

## v0.3.0

- Implement `__eq__` method for Quality.
- Fix `__eq__` method of Chord to support comparison between sharped and flatted chords.

## v0.2.9

- Implement `__eq__`, `__ne__`, `__setitem__` methods for ChordProgression.
- Implement `__eq__` method for Chord.

## v0.2.7

- Handle base note in Chord.components
- Contributor: @mstuttgart

## v0.2.6

- Enable setting scale on Chord.transpose
- Contributor: @jgvictores

## v0.2.5

- Refactor some classes not to modify instance variables.
- Update docstrings.

## v0.2.3

- Support 5th(power) chord.
- Add a utility to find chords from notes.

## v0.2.2

- Implement `__repr__` function.

## v0.2.1

- Support `__add__`, `__len__` and `__getitem__` functions in ChordProgression class.

## v0.2.0

- Add a class to handle chord progressions.

## v0.1.1

- Display flat or sharp by the scale.

## v0.1.0

- Add a function to get component notes of chord.


pychord/quality_list.py
# -*- coding: utf-8 -*-

QUALITY_DICT = {
    # chords consist of 3 notes
    '': [0, 4, 7],
    'maj': [0, 4, 7],
    'm': [0, 3, 7],
    'min': [0, 3, 7],
    'dim': [0, 3, 6],
    'arg': [0, 4, 8],
    'sus2': [0, 2, 7],
    'sus4': [0, 5, 7],
    # 4 notes
    '6': [0, 4, 7, 9],
    '7': [0, 4, 7, 10],
    '7-5': [0, 4, 6, 10],
    '7+5': [0, 4, 8, 10],
    '7sus4': [0, 5, 7, 10],
    'm7': [0, 3, 7, 10],
    'm7-5': [0, 3, 6, 10],
    'dim6': [0, 3, 6, 9],
    'M7': [0, 4, 7, 11],
    'M7+5': [0, 4, 8, 11],
    'mM7': [0, 3, 7, 11],
    'add9': [0, 4, 7, 14],
    '2': [0, 4, 7, 14],
    'add11': [0, 4, 7, 17],
    '4': [0, 4, 7, 17],
    # 5 notes
    '6/9': [0, 4, 7, 9, 14],
    '9': [0, 4, 7, 10, 14],
    'm9': [0, 3, 7, 10, 14],
    'M9': [0, 4, 7, 11, 14],
    '9sus4': [0, 5, 7, 10, 14],
    '7-9': [0, 4, 7, 10, 13],
    '7+9': [0, 4, 7, 10, 15],
    '11': [0, 7, 10, 14, 17],
    '7+11': [0, 4, 7, 10, 18],
    '7-13': [0, 4, 7, 10, 20],
    # 6 notes
    '13': [0, 4, 7, 10, 14, 21],
}
